read and write files using Pillow

mechanomi opened the pull request:
specifically:

- read files in using Pillow
- convert in-memory images to PNG and pass to cv2
- read cropped PNG images using Pillow
- write back out using the original file format

this is a backwards incompatible change. Pillow supports more formats than cv2
does, but it does not support all of the formats that cv2 does

public domain test image added. source:

https://en.wikipedia.org/wiki/File:Elizabeth_L._Remba_Gardner,_Women%27s_Airforce_Service_Pilots,_NARA-542191.jpg

closes https://github.com/leblancfg/autocrop/issues/45

leblancfg commented on autocrop/autocrop.py:
```diff
@@ -204,8 +245,17 @@ def main(input_d,
         reject_filename = os.path.join(reject_d, basename)
 
+        # Open the image and detect the file type
+        img_orig = Image.open(input_filename)
```
### Important
File I/O is typically the performance bottleneck in batch applications of autocrop. Since [OpenCV is usually an order of magnitude faster than PIL to read images](https://www.kaggle.com/vfdev5/pil-vs-opencv), it would make sense to use PIL only when reading image types OpenCV can't handle.

I know this means more code, but I don't think I'd merge this PR otherwise. Open to discussion , of course.

leblancfg commented on autocrop/autocrop.py:
```diff
@@ -226,8 +276,13 @@ def main(input_d,
+            # Encode the image as an in-memory PNG
+            img_png = cv2.imencode(".png", image)[1].tostring()
+            # Read the PNG data
+            img_new = Image.open(io.BytesIO(img_png))
```
### Important
The reason [the Windows tests fail](https://ci.appveyor.com/project/leblancfg/autocrop/builds/21742939/job/pgn8q8ah170dmax7#L193) is because the files aren't closed after reading them in. The Python idiom to read (and close) files is the following, called the context manager:

~~~python
with open(path) as file:
    img = Image(file)
~~~

As suggested previously, no need to decode to PNG, but please **do** use the context manager to read in the file with PIL. I strongly believe it should make the tests pass as they currently stand :smile:

mechanomi commented:
@leblancfg thanks for your review! I'll come back to this when I get a spare moment

mechanomi commented:
@leblancfg I believe I have addressed your comments now

the one thing I wasn't able to do was to remove the need to convert to a PNG in memory. I tried what you suggested but I can't get it work. can you help?

mechanomi commented:
note to self: the README needs updating too

also: I am not sure this is the best approach for performance. for any format not supported by OpenCV, we parse the image twice. that introduces lop-sided performance characteristics

mechanomi commented:
note to self: double space in `with  Image`

leblancfg commented:
@mechanomi see my inline comment re: using numpy to read in a PIL as array directly. Parsing is only done once in this case. 😄

mechanomi commented:
@leblancfg I tried using `np.asarray(img)` but it didn't work. unless there's a different comment that I'm missing?

when I get a bit of time, I'll give it another go

leblancfg commented:
Hey, good morning! :)

You can try the following:

~~~python
from PIL import Image
import numpy as np
import matplotlib.pyplot as plt


path = 'tests/data/remba-gardner.gif'

with open(path, 'rb') as f:
    img = Image.open(f)
    array = np.asarray(img)

plt.imshow(array)
~~~

mechanomi commented:
@leblancfg I believe I have fixed up all the outstanding issues

also, I've changed the main loop so that it picks the appropriate library to read the image, instead of trying and falling back to Pillow

I changed it to behave like this so that GIF files (or other files supported by Pillow) are not "second-class citizens" (incurring a performance hit of being read by OpenCV first)

leblancfg commented:
I haven’t forgotten this, should have the time to look into by the end of the week. Looks like flake8 failed on a code complexity issue &mdash; those are typically fast fixes.
